File: tools/geometry.py
"""Single source of truth for HUD geometry, in prototype pixel coordinates.

The design canvas is the prototype's native 1672x941.  Everything the build
consumes is derived from this file so the CSS/SVG and the verification tooling
can never drift apart.
"""

CANVAS = (1672, 941)

# Outer glass silhouette.  Not a rounded rectangle: a raised "title ear" on the
# left and a raised "tool pod" on the right fuse into the main body through
# S-fillets (a convex arc tangent to a concave arc), which border-radius cannot
# express -- hence the SVG path.
SHELL = {
    "left": 23,
    "right": 1648,
    "top": 411,          # main body top rim
    "bottom": 720,
    "r": 32,             # bottom corners
    "ear": {"right": 307, "top": 390, "r_tl": 29, "fillet": 34, "flat_end": 356},
    "pod": {"left": 1430, "top": 350, "r_tr": 45, "fillet": 54, "flat_start": 1333},
}

# Faint internal border delimiting the Status region; shares the shell's left,
# top and bottom edges, so only the right edge and its corner radius are real.
STATUS_PANE = {"right": 517, "bottom": 718, "r": 55}

# Height is 234 rather than the 233 the edge fit reports: the fit locates rim
# centres, and a 1.5px border drawn inside a 233px box puts the bottom rim's centre
# at 699.25 where the prototype has it at 700.5.
CARDS = {"top": 467, "height": 234, "width": 260, "pitch": 277, "first_left": 537, "r": 18}

TOOL_POD_BTN = {"d": 46, "cy": 391, "first_cx": 1439, "pitch": 77.5}

# Next-page disc sits on the right rim, half in the glass.  The live rim is
# gapped across this circle so the vertical highlight does not cut through it.
NEXT_BTN = {"cx": 1648, "cy": 566, "d": 40}

FAVOR_BAR = {"left": 95, "top": 666, "w": 232, "h": 14}
DETAIL_BTN = {"left": 419, "top": 654, "w": 74, "h": 30, "r": 15}

# Frames the content layer lays out inside.  These are the only content
# coordinates in the project: everything within a frame is placed by normal flow
# from the spacing and type scales in src/styles/tokens.css, so a new field or a
# fifth card costs a data entry and no measurement.
def regions():
    s = SHELL
    ear_top, pod = s["ear"]["top"], s["pod"]
    return {
        "status": {"x": s["left"], "y": ear_top, "w": STATUS_PANE["right"] - s["left"],
                   "h": STATUS_PANE["bottom"] - ear_top},
        "girls": {"x": STATUS_PANE["right"], "y": ear_top,
                  "w": s["right"] - STATUS_PANE["right"], "h": s["bottom"] - ear_top},
        # Height is set so the frame's centre line is the measured button centre,
        # which is what the row of buttons aligns to.
        "pod": {"x": pod["left"], "y": pod["top"], "w": s["right"] - pod["left"],
                "h": 2 * (TOOL_POD_BTN["cy"] - pod["top"])},
    }

# Measured optical constants (see tools/measure_optics.py output).
OPTICS = {
    "rim_peak": (218, 215, 230),      # brightest point of the top rim
    "body_left": (65, 70, 103),       # glass body colour, warm end
    "body_right": (47, 61, 102),      # glass body colour, cool end
    "top_glow_falloff": 40,           # px over which the inner top glow decays
    "bottom_bloom_rise": 85,          # px over which the inner bottom bloom rises
    "bottom_rim_lum": 250,
    "backdrop_sd_ratio": 0.11,        # local sd retained through the glass (~1/9)
}


def shell_path():
    """SVG path `d` for the outer glass silhouette, clockwise from the ear."""
    s = SHELL
    L, R, T, B, r = s["left"], s["right"], s["top"], s["bottom"], s["r"]
    ear, pod = s["ear"], s["pod"]
    eT, eR, eRad, eFil, eEnd = ear["top"], ear["right"], ear["r_tl"], ear["fillet"], ear["flat_end"]
    pL, pT, pRad, pFil, pStart = pod["left"], pod["top"], pod["r_tr"], pod["fillet"], pod["flat_start"]

    # S-fillet midpoints: the two tangent arcs meet halfway in both axes.
    ear_mid_x, ear_mid_y = (eR + eEnd) / 2, (eT + T) / 2
    pod_mid_x, pod_mid_y = (pStart + pL) / 2, (T + pT) / 2

    return " ".join([
        f"M {L + eRad} {eT}",
        f"L {eR} {eT}",
        f"A {eFil} {eFil} 0 0 1 {ear_mid_x} {ear_mid_y}",   # convex, curving down
        f"A {eFil} {eFil} 0 0 0 {eEnd} {T}",                 # concave, back to flat
        f"L {pStart} {T}",
        f"A {pFil} {pFil} 0 0 0 {pod_mid_x} {pod_mid_y}",    # concave, curving up
        f"A {pFil} {pFil} 0 0 1 {pL} {pT}",                  # convex, back to flat
        f"L {R - pRad} {pT}",
        f"A {pRad} {pRad} 0 0 1 {R} {pT + pRad}",            # pod top-right
        f"L {R} {B - r}",
        f"A {r} {r} 0 0 1 {R - r} {B}",                      # bottom-right
        f"L {L + r} {B}",
        f"A {r} {r} 0 0 1 {L} {B - r}",                      # bottom-left
        f"L {L} {eT + eRad}",
        f"A {eRad} {eRad} 0 0 1 {L + eRad} {eT}",            # ear top-left
        "Z",
    ])


def rim_segments():
    """Stroke pieces of the silhouette, each with its own measured colour.

    A single vertical gradient cannot express the real rim: the title-ear corner
    and the pod-rise S-curve are hot spots, the pod top is dimmer than the main
    top, the sides stay fairly bright (they are not the 'mid' of a y-gradient),
    and the whole bottom edge is near-white.  Each segment is a sub-path of
    shell_path() so the geometry cannot drift.
    """
    s = SHELL
    L, R, T, B, r = s["left"], s["right"], s["top"], s["bottom"], s["r"]
    ear, pod = s["ear"], s["pod"]
    eT, eR, eRad, eFil, eEnd = ear["top"], ear["right"], ear["r_tl"], ear["fillet"], ear["flat_end"]
    pL, pT, pRad, pFil, pStart = pod["left"], pod["top"], pod["r_tr"], pod["fillet"], pod["flat_start"]
    ear_mid_x, ear_mid_y = (eR + eEnd) / 2, (eT + T) / 2
    pod_mid_x, pod_mid_y = (pStart + pL) / 2, (T + pT) / 2

    return [
        {
            "id": "tl-ear",
            "d": f"M {L} {eT + eRad} A {eRad} {eRad} 0 0 1 {L + eRad} {eT} L {eR} {eT}",
            "color": "rgb(236,226,240)", "opacity": 0.96, "width": 2.15,
        },
        {
            "id": "main-top",
            "d": (f"M {eR} {eT} "
                  f"A {eFil} {eFil} 0 0 1 {ear_mid_x} {ear_mid_y} "
                  f"A {eFil} {eFil} 0 0 0 {eEnd} {T} "
                  f"L {pStart} {T}"),
            "color": "rgb(208,204,228)", "opacity": 0.90, "width": 2.0,
        },
        {
            "id": "pod-rise",
            "d": (f"M {pStart} {T} "
                  f"A {pFil} {pFil} 0 0 0 {pod_mid_x} {pod_mid_y} "
                  f"A {pFil} {pFil} 0 0 1 {pL} {pT}"),
            "color": "rgb(252,244,248)", "opacity": 1.0, "width": 2.35,
        },
        {
            "id": "pod-top",
            "d": f"M {pL} {pT} L {R - pRad} {pT}",
            "color": "rgb(198,176,184)", "opacity": 0.78, "width": 1.85,
        },
        {
            "id": "right",
            "d": f"M {R - pRad} {pT} A {pRad} {pRad} 0 0 1 {R} {pT + pRad} L {R} {B - r}",
            "color": "url(#rimRightGrad)", "opacity": 0.92, "width": 2.0,
        },
        {
            "id": "bottom",
            "d": f"M {R} {B - r} A {r} {r} 0 0 1 {R - r} {B} L {L + r} {B} A {r} {r} 0 0 1 {L} {B - r}",
            "color": "rgb(255,248,250)", "opacity": 1.0, "width": 2.25,
        },
        {
            "id": "left",
            "d": f"M {L} {B - r} L {L} {eT + eRad}",
            "color": "url(#rimLeftGrad)", "opacity": 0.90, "width": 2.0,
        },
    ]


def pod_floor_path():
    """Internal crease under the tool pod.  Ridge-traced at y 440 from the
    S-fillet (x 1336) to the pod's bottom-right quarter-circle, which meets the
    shell's right edge at the same radius as the top-right corner."""
    s = SHELL
    y = 440
    r = s["pod"]["r_tr"]
    x0 = s["pod"]["flat_start"] + 3
    x1 = s["right"] - r
    return f"M {x0} {y} L {x1} {y} A {r} {r} 0 0 1 {s['right']} {y - r}"


def status_pane_path():
    """Internal Status border: departs from the shell top, drops, returns to bottom."""
    s, p = SHELL, STATUS_PANE
    r = p["r"]
    return (f"M {p['right'] - r} {s['top']} "
            f"A {r} {r} 0 0 1 {p['right']} {s['top'] + r} "
            f"L {p['right']} {p['bottom'] - r} "
            f"A {r} {r} 0 0 1 {p['right'] - r} {p['bottom']}")


def convex_arcs():
    """The few silhouette arcs that read as a polished convex lens.

    A convex surface catches a highlight and a short milky fog just inside it --
    like a lightly ground convex mirror -- and nowhere else.  The open body of
    the glass stays clear; spreading haze along every border is what turned the
    panel into a frosted slab.
    """
    s = SHELL
    L, R, T, B, r = s["left"], s["right"], s["top"], s["bottom"], s["r"]
    ear, pod = s["ear"], s["pod"]
    eT, eR, eRad, eFil = ear["top"], ear["right"], ear["r_tl"], ear["fillet"]
    pL, pT, pRad, pFil, pStart = pod["left"], pod["top"], pod["r_tr"], pod["fillet"], pod["flat_start"]
    ear_mid_x, ear_mid_y = (eR + ear["flat_end"]) / 2, (eT + T) / 2
    pod_mid_x, pod_mid_y = (pStart + pL) / 2, (T + pT) / 2
    return [
        {
            "id": "ear-corner",
            "d": f"M {L} {eT + eRad} A {eRad} {eRad} 0 0 1 {L + eRad} {eT}",
            "color": "rgb(220, 228, 255)", "width": 14, "opacity": 0.10,
        },
        {
            "id": "ear-fillet",
            "d": f"M {eR} {eT} A {eFil} {eFil} 0 0 1 {ear_mid_x} {ear_mid_y}",
            "color": "rgb(210, 220, 255)", "width": 18, "opacity": 0.12,
        },
        {
            "id": "pod-fillet",
            "d": f"M {pod_mid_x} {pod_mid_y} A {pFil} {pFil} 0 0 1 {pL} {pT}",
            "color": "rgb(255, 232, 242)", "width": 20, "opacity": 0.14,
        },
        {
            "id": "pod-corner",
            "d": f"M {R - pRad} {pT} A {pRad} {pRad} 0 0 1 {R} {pT + pRad}",
            "color": "rgb(255, 236, 220)", "width": 14, "opacity": 0.08,
        },
    ]


def card_boxes():
    c = CARDS
    return [(c["first_left"] + i * c["pitch"], c["top"], c["width"], c["height"]) for i in range(4)]


if __name__ == "__main__":
    print("shell:", shell_path())
    print("\nstatus:", status_pane_path())
    print("\ncards:", card_boxes())

File: src/portrait/geometry.js
/* Portrait geometry.
   ------------------------------------------------------------------
   The landscape canvas is a pixel reproduction of a measured prototype, so its
   geometry is baked into geometry.json.  The portrait layout has exactly one fixed
   dimension -- the width.  Height is whatever the content needs, and it changes at
   runtime when the character preview expands downward.  A silhouette whose bottom
   edge is only known after layout cannot be baked, so the portrait shape is
   generated here from the same construction rules the landscape shell is built
   from:

     - a raised title ear on the top-left, joined to the main body by an S-fillet
       (a convex arc tangent to a concave arc), which border-radius cannot express
     - each panel's bottom edge carries the matching notch for the ear of the panel
       below it, offset outward by one seam width, so the gap between two panels is
       the same width all the way across.  Offset rule, as documented for the
       landscape dock: convex arcs grow by the gap, concave arcs shrink by it,
       centres unchanged.
     - the rim is painted as separate segments each carrying its own measured
       colour, not as one gradient along the outline

   Everything here is in portrait canvas units: the whole canvas is scaled by one
   factor k = containerWidth / PW, so one unit is one pixel of the portrait
   reference at container width 941.

   Scale note: the portrait reference draws type ~1.55x larger relative to its
   canvas than the landscape reference does relative to its own, because the same
   physical screen carries it at k ~ 0.41 instead of k ~ 1.  Rim widths, blur radii
   and frost grain carry the same RIM_K factor -- they are optical properties of the
   glass, so they scale with the drawing, not with the canvas. */

/** Design canvas width.  The layout is authored against this. */
export const PW = 941;

/* k has a floor, and the canvas width is what enforces it.
   ------------------------------------------------------------------
   Everything is scaled by k = containerWidth / canvasWidth, so the real size of the
   smallest type is 36 * k.  With the canvas pinned at 941 that lands at 13.8 real px
   on a 360px Android but only 12.2 on a 320px phone, below the 13px floor -- and a
   bigger number cannot fix it, because a narrower container always drags k back
   down.  So below the floor the canvas *narrows* instead: the same content reflows
   into fewer units, and k stops falling.

   Not symmetrical with a ceiling: a container wider than the design canvas is
   handled by capping k and centring the column, which is a stage concern, not a
   geometry one -- the canvas is still 941 units wide there. */
export const K_FLOOR = 13 / 36;

/** Canvas width for a given container width, never wider than the design canvas. */
export const canvasWidth = (containerWidth) =>
  Math.min(PW, Math.round(containerWidth / K_FLOOR));

/* Panel body.  `right` follows the canvas width so the two side margins stay equal
   at any width; everything else is an absolute unit and does not move. */
export const PANEL = { left: 33, right: PW - 33, r: 38 };

/** Panel edges for a canvas of `pw` units. */
export const panelEdges = (pw = PW) => ({ left: PANEL.left, right: pw - PANEL.left, r: PANEL.r });

/* Raised title tab on the top-left of every panel.  Width is set by the longest
   script title ("Profile") at the portrait script size, and it has to leave the
   Status panel's top edge a flat run between its own fillet and the pod's:
   ear ends at 33 + 280 + 91 = 404, the pod's fillet starts at 461, so 57 units. */
export const EAR = { width: 280, r: 34, drop: 44, fillet: 58 };

/** Raised tool tab on the top-right -- only the Status panel carries one. */
export const POD = { width: 356, r: 46 };

/* Tool ring buttons inside the pod.  Four 76-unit rings (plus 12 gaps) fill the
   356-unit pod the way three 96-unit rings used to, leaving the same ~16 units of
   side slack so the outer two sit inside the pod's corner radius.  76 units is
   31.5 real px at a 390px-wide container; the hit area is still widened to 44 by
   a pseudo-element rather than by growing the ring.

   `count` is a ceiling and not a knob.  Five rings need 428 units, and the pod has
   only 57 units of flat top edge to grow into before its fillet meets the title
   ear's -- so the fifth does not fit even after spending all of it.  Shrinking the
   ring to make five fit is worse than tight: the 44px hit areas already overlap at
   this pitch (3.8 real px at a 430-wide container, 12.2 at 320), and a 58-unit ring
   would put that overlap past half the button.  So the pod holds only the four rings the
   player presses without reading them, and every other destination is an entry in
   `destinations` (src/data.js), drawn as a labelled cell in the panel body -- see
   portraitDestGrid in portrait/content.js.  New destinations go there, not here. */
export const TOOL = { d: 76, gap: 12, count: 4 };

/** One seam width, used for every gap between stacked panels. */
export const SEAM = 26;

/** Portrait draws at 1.55x the landscape scale; optical widths follow. */
export const RIM_K = 1.55;

/* Cards in the Girls rail.
   The landscape card is landscape-proportioned (260x234, art left, text in a glass
   pane on the right) because it lives in a 1131-wide rail.  Reusing that proportion
   here cost 596 units of width to show 1.28 cards, and left a 296-wide text column
   holding 32-unit type -- sparse and awkward.  The portrait card rotates the same
   idea instead: art on top, a vertical fade, the text pane along the bottom.  At
   330 the rail shows ~2.3 cards, so it reads as a rail rather than a single slide. */
export const CARD = { w: 330, h: 574, gap: 24, r: 30 };

/**
 * Horizontal run of an S-fillet, and the point where its two arcs meet.
 *
 * For the arcs to be tangent to each other and to both horizontals the centres
 * must be exactly r1 + r2 apart, which forces
 *     dx^2 = drop * (2 * (r1 + r2) - drop)
 * Because (F + g) + (F - g) = 2F, dx does not depend on the gap -- which is why a
 * notch spans exactly the same x range as the ear it mates with, and therefore why
 * the seam keeps one width.  For equal radii the join is simply the midpoint in
 * both axes (the landscape pod rise: 1381.5, 380.5).  Checked against the shell ear
 * and the dock notch by tools/check_offset_rule.py.
 */
export function sFillet(drop, r1, r2) {
  const s = r1 + r2;
  const dx = Math.sqrt(drop * (2 * s - drop));
  const t = r1 / s;
  return { dx, jx: t * dx, jy: r1 - t * (r1 + r2 - drop) };
}

const EAR_F = sFillet(EAR.drop, EAR.fillet, EAR.fillet);
/** Horizontal run of a panel's own ear fillet -- gap-independent, so shared. */
export const FILLET_RUN = EAR_F.dx;

/* Key x-coordinates of a panel's top edge, shared by the path, the rim and the
   notch above it, so none of the three can drift.  The ear is anchored to the left
   edge and so is width-independent; the pod hangs off the right edge. */
function topEdge(pod, pw = PW) {
  const L = PANEL.left;
  const earRight = L + EAR.width;
  const earEnd = earRight + FILLET_RUN;
  if (!pod) return { earRight, earEnd, podStart: null, podLeft: null };
  const podLeft = pw - L - POD.width;
  return { earRight, earEnd, podStart: podLeft - FILLET_RUN, podLeft };
}

/** Left edge of the tool pod's button row, for the stylesheet. */
export const podRowLeft = (pw = PW) =>
  pw - PANEL.left - POD.width
  + (POD.width - (TOOL.count * TOOL.d + (TOOL.count - 1) * TOOL.gap)) / 2;

/* A panel's own ear fillet: steps down from the ear top to the main top.
   Equal radii, so the join is the midpoint. */
const earFall = (earRight, eT) =>
  `A ${EAR.fillet} ${EAR.fillet} 0 0 1 ${earRight + FILLET_RUN / 2} ${eT + EAR.drop / 2} ` +
  `A ${EAR.fillet} ${EAR.fillet} 0 0 0 ${earRight + FILLET_RUN} ${eT + EAR.drop}`;

/* The pod fillet: the same shape rotated 180 degrees -- steps up from the main
   top to the pod top. */
const podRise = (podStart, mT) =>
  `A ${EAR.fillet} ${EAR.fillet} 0 0 0 ${podStart + FILLET_RUN / 2} ${mT - EAR.drop / 2} ` +
  `A ${EAR.fillet} ${EAR.fillet} 0 0 1 ${podStart + FILLET_RUN} ${mT - EAR.drop}`;

/* The notch in a panel's bottom edge, traversed right to left because the path
   runs clockwise.  Reversing an arc flips its sweep flag, so the notch's arcs come
   in the opposite order and with opposite sweeps to the ear they mirror.  Radii
   follow the offset rule: the convex arc grows by the seam, the concave one shrinks
   by it, and the two bottom-left corners facing across the seam differ by it too. */
function notchRTL(nextEar) {
  const bottomEar = nextEar - SEAM;
  /* The notch mirrors the ear below, which is anchored to the left edge, so it does
     not depend on the canvas width. */
  const { earRight, earEnd } = topEdge(false);
  const rIn = EAR.fillet + SEAM;
  const rOut = EAR.fillet - SEAM;
  const { jx, jy } = sFillet(EAR.drop, rIn, rOut);
  const rBL = EAR.r - SEAM;
  return {
    bottomEar,
    rBL,
    d: [
      `L ${earEnd} ${bottomEar + EAR.drop}`,
      `A ${rOut} ${rOut} 0 0 1 ${earRight + jx} ${bottomEar + jy}`,
      `A ${rIn} ${rIn} 0 0 0 ${earRight} ${bottomEar}`,
      `L ${PANEL.left + rBL} ${bottomEar}`,
      `A ${rBL} ${rBL} 0 0 1 ${PANEL.left} ${bottomEar - rBL}`,
    ].join(' '),
  };
}

/* The bottom edge right to left, flat or notched.  `bottom` is the main bottom
   level; with a notch the level under the ear rises to nextEar - SEAM, and the main
   level must equal that plus the drop so the seam stays uniform. */
function bottomEdgeRTL(bottom, nextEar, pw = PW) {
  const { left: L, right: R, r } = panelEdges(pw);
  if (nextEar == null) {
    return `A ${r} ${r} 0 0 1 ${R - r} ${bottom} L ${L + r} ${bottom} A ${r} ${r} 0 0 1 ${L} ${bottom - r}`;
  }
  return `A ${r} ${r} 0 0 1 ${R - r} ${bottom} ${notchRTL(nextEar).d}`;
}

/** Where a notched panel's bottom must sit so the seam is one width. */
export const notchedBottom = (nextEar) => nextEar - SEAM + EAR.drop;

/** Where a notched panel's left rim has to stop. */
const leftRimStop = (bottom, nextEar) =>
  nextEar == null ? bottom - PANEL.r : nextEar - SEAM - (EAR.r - SEAM);

/**
 * Outer silhouette of one panel.
 * `earTop` is the highest point (the raised tab); the main body top sits EAR.drop
 * below it.  `bottom` is the main bottom level, from measured content height.
 * `nextEar` is the ear top of the panel below, or null for the last panel.
 */
export function panelPath({ earTop: eT, bottom, pod = false, nextEar = null, pw = PW }) {
  const { left: L, right: R, r } = panelEdges(pw);
  const mT = eT + EAR.drop;
  const { earRight, earEnd, podStart, podLeft } = topEdge(pod, pw);

  const head = pod
    ? `L ${podStart} ${mT} ${podRise(podStart, mT)} L ${R - POD.r} ${eT} ` +
      `A ${POD.r} ${POD.r} 0 0 1 ${R} ${eT + POD.r}`
    : `L ${R - r} ${mT} A ${r} ${r} 0 0 1 ${R} ${mT + r}`;

  return [
    `M ${L + EAR.r} ${eT}`,
    `L ${earRight} ${eT}`,
    earFall(earRight, eT),
    head,
    `L ${R} ${bottom - r}`,
    bottomEdgeRTL(bottom, nextEar, pw),
    `L ${L} ${eT + EAR.r}`,
    `A ${EAR.r} ${EAR.r} 0 0 1 ${L + EAR.r} ${eT}`,
    'Z',
  ].join(' ');
}

/* Rim segments.  Colours are the landscape shell's measured per-edge values --
   sampled per edge and per axis, not per composition, so a portrait panel's top
   edge is the same light as a landscape top edge.  Widths carry RIM_K.  The side
   rims use objectBoundingBox gradients so one definition serves any panel height. */
const w = (n) => +(n * RIM_K).toFixed(2);

export function panelRim({ earTop: eT, bottom, pod = false, nextEar = null, pw = PW }) {
  const { left: L, right: R, r } = panelEdges(pw);
  const mT = eT + EAR.drop;
  const { earRight, earEnd, podStart, podLeft } = topEdge(pod, pw);

  const segs = [
    {
      id: 'ear-top',
      d: `M ${L} ${eT + EAR.r} A ${EAR.r} ${EAR.r} 0 0 1 ${L + EAR.r} ${eT} L ${earRight} ${eT}`,
      color: 'rgb(236,226,240)', opacity: 0.96, width: w(2.15),
    },
    {
      id: 'ear-fall',
      d: `M ${earRight} ${eT} ${earFall(earRight, eT)}`,
      color: 'rgb(250,242,248)', opacity: 1.0, width: w(2.3),
    },
    {
      id: 'main-top',
      d: `M ${earEnd} ${mT} L ${pod ? podStart : R - r} ${mT}`,
      color: 'rgb(208,204,228)', opacity: 0.90, width: w(2.0),
    },
  ];

  if (pod) {
    segs.push({
      id: 'pod-rise',
      d: `M ${podStart} ${mT} ${podRise(podStart, mT)}`,
      color: 'rgb(252,244,248)', opacity: 1.0, width: w(2.35),
    }, {
      id: 'pod-top',
      d: `M ${podLeft} ${eT} L ${R - POD.r} ${eT}`,
      color: 'rgb(198,176,184)', opacity: 0.78, width: w(1.85),
    }, {
      id: 'right',
      d: `M ${R - POD.r} ${eT} A ${POD.r} ${POD.r} 0 0 1 ${R} ${eT + POD.r} L ${R} ${bottom - r}`,
      color: 'url(#pRimRight)', opacity: 0.92, width: w(2.0),
    });
  } else {
    segs.push({
      id: 'right',
      d: `M ${R - r} ${mT} A ${r} ${r} 0 0 1 ${R} ${mT + r} L ${R} ${bottom - r}`,
      color: 'url(#pRimRight)', opacity: 0.92, width: w(2.0),
    });
  }

  /* Every bottom edge of the landscape shell is near-white; the notch rides along
     on the same stroke so the seam reads as one continuous edge. */
  segs.push({
    id: 'bottom',
    d: `M ${R} ${bottom - r} ${bottomEdgeRTL(bottom, nextEar, pw)}`,
    color: 'rgb(255,248,250)', opacity: 1.0, width: w(2.25),
  }, {
    id: 'left',
    d: `M ${L} ${leftRimStop(bottom, nextEar)} L ${L} ${eT + EAR.r}`,
    color: 'url(#pRimLeft)', opacity: 0.90, width: w(2.0),
  });
  return segs;
}

/* The arcs that read as a polished convex lens: the ear's own corner, the convex
   halves of the S-fillets, and the pod corner.  Same restraint as landscape --
   the open body stays clear, only convex surfaces scatter. */
export function panelConvexArcs({ earTop: eT, pod = false, pw = PW }) {
  const { left: L, right: R } = panelEdges(pw);
  const F = EAR.fillet;
  const { earRight, podStart, podLeft } = topEdge(pod, pw);

  const arcs = [
    {
      id: 'ear-corner',
      d: `M ${L} ${eT + EAR.r} A ${EAR.r} ${EAR.r} 0 0 1 ${L + EAR.r} ${eT}`,
      color: 'rgb(220, 228, 255)', width: w(14), opacity: 0.10,
    },
    {
      id: 'ear-fillet',
      d: `M ${earRight} ${eT} A ${F} ${F} 0 0 1 ${earRight + FILLET_RUN / 2} ${eT + EAR.drop / 2}`,
      color: 'rgb(210, 220, 255)', width: w(18), opacity: 0.12,
    },
  ];
  if (pod) {
    arcs.push({
      id: 'pod-fillet',
      d: `M ${podStart + FILLET_RUN / 2} ${eT + EAR.drop / 2} A ${F} ${F} 0 0 1 ${podLeft} ${eT}`,
      color: 'rgb(255, 232, 242)', width: w(20), opacity: 0.14,
    }, {
      id: 'pod-corner',
      d: `M ${R - POD.r} ${eT} A ${POD.r} ${POD.r} 0 0 1 ${R} ${eT + POD.r}`,
      color: 'rgb(255, 236, 220)', width: w(14), opacity: 0.08,
    });
  }
  return arcs;
}
